relayer/processor/path_processor.go

```diff
@@ -169,32 +180,71 @@ func (pp *PathProcessor) HandleNewData(chainID string, cacheData ChainProcessorC
 }
 
 // Run executes the main path process.
-func (pp *PathProcessor) Run(ctx context.Context) {
+func (pp *PathProcessor) Run(ctx context.Context, ctxCancel func(), messageLifecycle MessageLifecycle) {
+	var retryTimer *time.Timer
 	for {
+		// block until we have any signals to process
 		select {
 		case <-ctx.Done():
+			pp.log.Debug("Context done, quitting PathProcessor",
+				zap.String("chain_id_1", pp.pathEnd1.info.ChainID),
+				zap.String("chain_id_2", pp.pathEnd2.info.ChainID),
+				zap.String("client_id_1", pp.pathEnd1.info.ClientID),
+				zap.String("client_id_2", pp.pathEnd2.info.ClientID),
+				zap.Error(ctx.Err()),
+			)
 			return
-
 		case d := <-pp.pathEnd1.incomingCacheData:
 			// we have new data from ChainProcessor for pathEnd1
-			pp.pathEnd1.MergeCacheData(d)
+			pp.pathEnd1.MergeCacheData(ctx, ctxCancel, d, messageLifecycle)
 
 		case d := <-pp.pathEnd2.incomingCacheData:
 			// we have new data from ChainProcessor for pathEnd2
-			pp.pathEnd2.MergeCacheData(d)
+			pp.pathEnd2.MergeCacheData(ctx, ctxCancel, d, messageLifecycle)
 
 		case <-pp.retryProcess:
 			// No new data to merge in, just retry handling.
 		}
 
+		// Fully flush pathEnd incoming data before processing
+		for len(pp.pathEnd1.incomingCacheData) > 0 {
+			pp.pathEnd1.MergeCacheData(ctx, ctxCancel, <-pp.pathEnd1.incomingCacheData, messageLifecycle)
+		}
+		for len(pp.pathEnd2.incomingCacheData) > 0 {
+			pp.pathEnd2.MergeCacheData(ctx, ctxCancel, <-pp.pathEnd2.incomingCacheData, messageLifecycle)
+		}
```

Comment:
I don't have enough context to tell whether it is okay to read from these channels in sequence, as opposed to 

```go
for len(p1)>0 || len(p2)>0 {
  select {
    case x := <-ch1:
      p1.merge(x)
    case x := <-ch2:
      p2.merge(x)
    case <-ctx.Done():
      // ...
  }
}
```

Reply:
Since the `select` chooses a randomly available signal, I want to make sure all channels are flushed before moving to the `processLatestMessages`, since I was seeing it run without the most recent information that had been pushed to the channels due to the amount of time that the previous `processLatestMessages` cycle took.

Reply:
This `select` example will work well. It looks about identical to the one above, so I consolidated them into a single function `processAvailableSignals`. The above one will block until signals are available, and the one after that can process any additional available signals before moving on to the `processLatestMessages`.